File: app/src/main/java/com/whiteside/insta/VisitProfile.java
package com.whiteside.insta;

import android.content.Intent;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QuerySnapshot;
import com.whiteside.insta.databinding.ActivityVisitProfileBinding;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import Controller.PostsRecyclerViewAdapter;
import Model.Post;
import Model.Profile;

import static android.content.ContentValues.TAG;

public class VisitProfile extends AppCompatActivity {


    private String UID;
    private FirebaseAuth auth;

    private RecyclerView recyclerView;
    private PostsRecyclerViewAdapter adapter;
    private Map<String, Post> posts;
    private List<Post> list;

    private FirebaseFirestore fStore;
    private Profile profile;
    private ActivityVisitProfileBinding bind;

    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        bind = ActivityVisitProfileBinding.inflate(getLayoutInflater());
        setContentView(bind.getRoot());

        profile = new Profile();
        list = new ArrayList<>();
        auth = FirebaseAuth.getInstance();
        fStore = FirebaseFirestore.getInstance();

        recyclerView = findViewById(R.id.profile_posts);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        posts = new HashMap();
        adapter = new PostsRecyclerViewAdapter(list, this);
        recyclerView.setAdapter(adapter);

        Intent intent = getIntent();
        Bundle bundle = intent.getExtras();
        UID = bundle.get("UID").toString();

        getInfo();
        getProfilePostsFromFirebaseFireStore();
    }

    @Override
    protected void onStart() {
        super.onStart();

    }

    private void getInfo() {
        fStore.collection("Users")
                .document(UID)
                .get()
                .addOnSuccessListener(this, new OnSuccessListener<DocumentSnapshot>() {
                    @Override
                    public void onSuccess(DocumentSnapshot ds) {
                        profile = ds.toObject(Profile.class);
                        bind.userProfileName.setText(profile.getName());
                        bind.userProfileShortBio.setText(profile.getBio());
                        bind.userProfilePhoto.setImageBitmap(BitmapFactory.decodeByteArray(
                                profile.getProfilePic().toBytes(), 0, profile.getProfilePic().toBytes().length));

                        if (profile.getFriends().contains(auth.getUid())) {
                            bind.addFriend.setVisibility(View.INVISIBLE);
                        } else if (UID.equals(auth.getUid())) {
                            bind.addFriend.setVisibility(View.INVISIBLE);
                        } else {
                            bind.addFriend.setVisibility(View.VISIBLE);
                        }
                    }
                });


    }

    private void getProfilePostsFromFirebaseFireStore() {
        fStore.collection("Users")
                .document(UID)
                .collection("Posts")
                .get()
                .addOnCompleteListener(this, new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        if (!task.isSuccessful()) {
                            Log.e(TAG, "Error happened while download posts");
                            Log.e(TAG, task.getException().getStackTrace().toString());
                        } else {
                            if (!task.getResult().isEmpty()) {
                                for (DocumentSnapshot ds : task.getResult().getDocuments()) {
                                    Post p = ds.toObject(Post.class);
                                    posts.put(p.getID(), p);

                                    list = new ArrayList<>();
                                    list.addAll(posts.values());
                                    Collections.sort(list);
                                    adapter.update(list);
                                }
                            } else {
                                Toast.makeText(VisitProfile.this, "No Posts", Toast.LENGTH_SHORT).show();
                            }
                        }
                    }
                });
    }

    public void addFriendClicked(View view) {
        profile.addFriendRequest(auth.getUid());

        FirebaseFirestore.getInstance()
                .collection("Users")
                .document(Objects.requireNonNull(UID))
                .update("friendRequests", profile.getFriendRequests())
                .addOnCompleteListener(new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {

                    }
                });
    }
}

File: app/src/main/java/com/whiteside/insta/VisitProfile.kt
package com.whiteside.insta

import Controller.PostsRecyclerViewAdapter
import Model.Post
import Model.Profile
import android.content.ContentValues
import android.graphics.BitmapFactory
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.whiteside.insta.databinding.ActivityVisitProfileBinding
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.HashMap

class VisitProfile : AppCompatActivity() {
    private var UID: String? = null
    private var auth: FirebaseAuth? = FirebaseAuth.getInstance()
    private var recyclerView: RecyclerView? = null
    private var adapter: PostsRecyclerViewAdapter? = null
    private var posts: HashMap<String, Post> = HashMap()
    private var list: ArrayList<Post> = ArrayList()
    private var fStore: FirebaseFirestore? = FirebaseFirestore.getInstance()
    private var profile: Profile? = Profile()
    private var bind: ActivityVisitProfileBinding? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        bind = ActivityVisitProfileBinding.inflate(layoutInflater)
        setContentView(bind!!.root)

        adapter = PostsRecyclerViewAdapter(list, this)

        recyclerView = bind!!.profilePosts
        recyclerView!!.setHasFixedSize(true)
        recyclerView!!.setLayoutManager(LinearLayoutManager(this))
        recyclerView!!.setAdapter(adapter)

        val bundle = intent.extras
        UID = bundle!!["UID"].toString()
        info

        profilePostsFromFirebaseFireStore
    }

    override fun onStart() {
        super.onStart()
    }

    private val info: Unit
        private get() {
            fStore!!.collection("Users")
                    .document(UID!!)
                    .get()
                    .addOnSuccessListener(this) { ds ->
                        profile = ds.toObject(Profile::class.java)
                        bind!!.userProfileName.text = profile!!.name
                        bind!!.userProfileShortBio.text = profile!!.bio
                        bind!!.userProfilePhoto.setImageBitmap(BitmapFactory.decodeByteArray(
                                profile!!.profilePic!!.toBytes(), 0, profile!!.profilePic!!.toBytes().size))
                        if (profile!!.getFriends()!!.contains(auth!!.uid!!)) {
                            bind!!.addFriend.visibility = View.INVISIBLE
                        } else if (UID == auth!!.uid) {
                            bind!!.addFriend.visibility = View.INVISIBLE
                        } else {
                            bind!!.addFriend.visibility = View.VISIBLE
                        }
                    }
        }
    private val profilePostsFromFirebaseFireStore: Unit
        private get() {
            fStore!!.collection("Users")
                    .document(UID!!)
                    .collection("Posts")
                    .get()
                    .addOnCompleteListener(this) { task ->
                        if (!task.isSuccessful) {
                            Log.e(ContentValues.TAG, "Error happened while download posts")
                            Log.e(ContentValues.TAG, task.exception!!.stackTrace.toString())
                        } else {
                            if (!task.result!!.isEmpty) {
                                for (ds in task.result!!.documents) {
                                    val p = ds.toObject(Post::class.java)
                                    posts.put(p!!.Id!!,  p)
                                    list = ArrayList()
                                    list.addAll(posts!!.values)
                                    Collections.sort(list)
                                    adapter!!.update(list)
                                }
                            } else {
                                Toast.makeText(this@VisitProfile, "No Posts", Toast.LENGTH_SHORT).show()
                            }
                        }
                    }
        }

    fun addFriendClicked(view: View?) {
        profile!!.addFriendRequest(auth!!.uid!!)
        FirebaseFirestore.getInstance()
                .collection("Users")
                .document(UID!!)
                .update("friendRequests", profile!!.getFriendRequests())
                .addOnCompleteListener { }
    }
}